src/vs/code/electron-main/menubar.ts

```diff
@@ -495,22 +474,57 @@ export class Menubar {
 		}
 	}
 
-	private setMenuById(menu: Electron.Menu, menuId: string): void {
-		console.log(`Attempting to set menu for ${menuId}`);
-
-		// Build dynamic menu
-		this.menubarMenus[menuId].items.forEach((item: IMenubarMenuItemAction | IMenubarMenuItemSeparator) => {
-			if (item.id === 'vscode.menubar.separator') {
+	private setMenu(menu: Electron.Menu, items: Array<MenubarMenuItem>) {
+		items.forEach((item: MenubarMenuItem) => {
+			if (isMenubarMenuItemSeparator(item)) {
 				menu.append(__separator__());
-			} else {
-				let menuItem: Electron.MenuItem;
-				let action: IMenubarMenuItemAction = <IMenubarMenuItemAction>item;
-				menuItem = this.createMenuItem(action.label, action.id, action.enabled, action.checked);
+			} else if (isMenubarMenuItemSubmenu(item)) {
+				const submenu = new Menu();
+				const submenuItem = new MenuItem({ label: this.mnemonicLabel(item.label), submenu: submenu });
+				this.setMenu(submenu, item.submenu.items);
+				menu.append(submenuItem);
+			} else if (isMenubarMenuItemAction(item)) {
+				if (item.id === 'workbench.action.openRecent') {
+					this.insertRecentMenuItems(menu);
+				}
+
+				// Store the keybinding
+				if (item.keybinding) {
+					this.keybindings[item.id] = item.keybinding;
+				}
+
+				const menuItem = this.createMenuItem(item.label, item.id, item.enabled, item.checked);
 				menu.append(menuItem);
 			}
 		});
 	}
 
+	private setMenuById(menu: Electron.Menu, menuId: string): void {
+		this.setMenu(menu, this.menubarMenus[menuId].items);
+	}
+
+	private insertRecentMenuItems(menu: Electron.Menu) {
```

Comment:
Feels like it deos not belong in this file. this hsould be contributed by the filesActions.contributions imho

Reply:
same as above?